GH-Actions (todo) : Add `-Og` and `gcc-12`

- https://github.com/Cyan4973/xxHash/pull/804#issuecomment-1447244510
- https://github.com/Cyan4973/xxHash/issues/800

`v0.8.1` and [`HEAD`](https://github.com/Cyan4973/xxHash/tree/2b328a10983d232364ceda15df1d33531b5f0eb5) may have issue which is not related to `inline` problems.  But `-Og` helps us to detect it.

```
git clone https://github.com/Cyan4973/xxHash.git xxHash-v0.8.1 --branch v0.8.1 --single-branch --depth 1
cd xxHash-v0.8.1

make clean all
# OK

MOREFLAGS="-Og" make clean all
# NG (warnings)
```

<details><summary>Here's actual warning messages.</summary>

```
cc -O3 -Wall -Wextra -Wconversion -Wcast-qual -Wcast-align -Wshadow -Wstrict-aliasing=1 -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wundef -Wpointer-arith -Wformat-security -Wvla -Wformat=2 -Winit-self -Wfloat-equal -Wwrite-strings -Wredundant-decls -Wstrict-overflow=2 -Og -DXXH_INLINE_ALL cli/xxhsum.c -o xxhsum_inlinedXXH
In file included from cli/xsum_sanity_check.c:31,
                 from cli/xxhsum.c:41:
In function ‘XXH_INLINE_XXH3_64bits_reset_withSeed’,
    inlined from ‘XXH_INLINE_XXH3_128bits_reset_withSeed’ at cli/../xxhash.h:5402:12,
    inlined from ‘localXXH128_stream_seeded’ at cli/xsum_bench.c:176:5:
cli/../xxhash.h:4733:46: warning: ‘state.extSecret’ may be used uninitialized [-Wmaybe-uninitialized]
 4733 |     if ((seed != statePtr->seed) || (statePtr->extSecret != NULL))
      |                                      ~~~~~~~~^~~~~~~~~~~
In file included from cli/xxhsum.c:42:
cli/xsum_bench.c: In function ‘localXXH128_stream_seeded’:
cli/xsum_bench.c:174:18: note: ‘state’ declared here
  174 |     XXH3_state_t state;
      |                  ^~~~~
In file included from cli/xsum_sanity_check.c:31,
                 from cli/xxhsum.c:41:
In function ‘XXH_INLINE_XXH3_64bits_reset_withSeed’,
    inlined from ‘localXXH3_stream_seeded’ at cli/xsum_bench.c:160:5:
cli/../xxhash.h:4733:46: warning: ‘state.extSecret’ may be used uninitialized [-Wmaybe-uninitialized]
 4733 |     if ((seed != statePtr->seed) || (statePtr->extSecret != NULL))
      |                                      ~~~~~~~~^~~~~~~~~~~
In file included from cli/xxhsum.c:42:
cli/xsum_bench.c: In function ‘localXXH3_stream_seeded’:
cli/xsum_bench.c:158:18: note: ‘state’ declared here
  158 |     XXH3_state_t state;
      |                  ^~~~~
```

</details>

It warns that at xxhash.h#L4733, if `(statePtr != NULL) && (seed != 0) && !(seed != seedPtr->seed)`, we may read uninitialized variable `statePtr->extSecret`.

https://github.com/Cyan4973/xxHash/blob/v0.8.1/xxhash.h#L4733

```C
        XXH_PUBLIC_API XXH_errorcode
        XXH3_64bits_reset_withSeed(XXH3_state_t* statePtr, XXH64_hash_t seed)
        {
            if (statePtr == NULL) return XXH_ERROR;
            if (seed==0) return XXH3_64bits_reset(statePtr);
/*4733*/    if ((seed != statePtr->seed) || (statePtr->extSecret != NULL))
                XXH3_initCustomSecret(statePtr->customSecret, seed);
            XXH3_reset_internal(statePtr, seed, NULL, XXH_SECRET_DEFAULT_SIZE);
            return XXH_OK;
        }
```

Our initizliation is done by `XXH3_INITSTATE()`.  Note that we don't have any explicit initializer list for `state`.

https://github.com/Cyan4973/xxHash/blob/v0.8.1/cli/xsum_bench.c#L158

```C
        static XSUM_U32 localXXH3_stream_seeded(const void* buffer, size_t bufferSize, XSUM_U32 seed)
        {
/*158*/     XXH3_state_t state;
            XXH3_INITSTATE(&state);
            XXH3_64bits_reset_withSeed(&state, (XXH64_hash_t)seed);
            XXH3_64bits_update(&state, buffer, bufferSize);
            return (XSUM_U32)XXH3_64bits_digest(&state);
        }
```

`XXH3_INITSTATE()` only set `statePtr->seed`.  But it doesn't touch `statePtr->extSecret`.

https://github.com/Cyan4973/xxHash/blob/v0.8.1/xxhash.h#L1113

```
#define XXH3_INITSTATE(XXH3_state_ptr)   { (XXH3_state_ptr)->seed = 0; }
```

I think adding `(XXH3_state_ptr)->extSecret = NULL;` to `XXH3_INITSTATE()` is the most easy and safe fix for users who doesn't care about internals of XXH3.
But not sure about benchmarks.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -73,6 +73,7 @@ jobs:
 
           # gcc
           { pkgs: '',                                  cc: gcc,       cxx: g++,         avx512: 'true',  os: ubuntu-latest, },
+          { pkgs: 'gcc-12  g++-12  lib32gcc-12-dev',   cc: gcc-12,    cxx: g++-12,      avx512: 'true',  os: ubuntu-22.04,  },
           { pkgs: 'gcc-11  g++-11  lib32gcc-11-dev',   cc: gcc-11,    cxx: g++-11,      avx512: 'true',  os: ubuntu-20.04,  },
           { pkgs: 'gcc-10  g++-10  lib32gcc-10-dev',   cc: gcc-10,    cxx: g++-10,      avx512: 'true',  os: ubuntu-20.04,  },
           { pkgs: 'gcc-9   g++-9   lib32gcc-9-dev',    cc: gcc-9,     cxx: g++-9,       avx512: 'true',  os: ubuntu-20.04,  },

--- a/Makefile
+++ b/Makefile
@@ -316,6 +316,11 @@ clangtest: clean
 	@echo ---- test clang compilation ----
 	CC=clang MOREFLAGS="-Werror -Wconversion -Wno-sign-conversion" $(MAKE) all
 
+.PHONY: gcc-og-test
+gcc-og-test: clean
+	@echo ---- test gcc -Og compilation ----
+	CFLAGS="-Og -Wall -Wextra -Wundef -Wshadow -Wcast-align -Werror -fPIC" MOREFLAGS="-Werror" $(MAKE) all
+
 .PHONY: cxxtest
 cxxtest: clean
 	@echo ---- test C++ compilation ----
@@ -420,7 +425,7 @@ test-inline:
 
 .PHONY: test-all
 test-all: CFLAGS += -Werror
-test-all: test test32 test-unicode clangtest cxxtest usan test-inline listL120 trailingWhitespace test-xxh-nnn-sums
+test-all: test test32 test-unicode clangtest gcc-og-test cxxtest usan test-inline listL120 trailingWhitespace test-xxh-nnn-sums
 
 .PHONY: test-tools
 test-tools:

--- a/xxhash.h
+++ b/xxhash.h
@@ -1285,7 +1285,12 @@ struct XXH3_state_s {
  * Note that this doesn't prepare the state for a streaming operation,
  * it's still necessary to use XXH3_NNbits_reset*() afterwards.
  */
-#define XXH3_INITSTATE(XXH3_state_ptr)   { (XXH3_state_ptr)->seed = 0; }
+#define XXH3_INITSTATE(XXH3_state_ptr)                       \
+    do {                                                     \
+        XXH3_state_t* tmp_xxh3_state_ptr = (XXH3_state_ptr); \
+        tmp_xxh3_state_ptr->seed = 0;                        \
+        tmp_xxh3_state_ptr->extSecret = NULL;                \
+    } while(0)
 
 
 /*!

